### PR TITLE
added eager tutorial and attacks.

### PR DESCRIPTION
**Pull request by feedforward**
For eager execution. 
1. A subclass of cleverhans.model.Model for tf.eager models is added. (cleverhans.model.Model_Eager), 
this uses only tf_eager without any keras dependencies.

2. New file attacks_eager is added, (existing attacks run through session using feed_dict making it too hard to adapt.), 
but structure of attacks.py is adapted to attacks_eager.py,  so objects of both attack classes can be replaced with out any difference.

3. tutorial to show the use tutorial_mnist_eager.py

4. unit tests for all the above.

### CODE REVIEW

**Comment by npapernot** [comment]
@nottombrown @david-berthelot , would you be able to review this PR since you use this logic more often than I do?

**Comment by nottombrown** [comment]
Hmmm, I haven't actually used `tf.eager` for a project yet. @david-berthelot have you?

File: cleverhans/model_eager.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+
+if LooseVersion(tf.__version__) < LooseVersion('1.8.0'):
+    error = ('For eager execution',
+             'use Tensorflow version greather than 1.8.0.')
+    print(error)
+    exit(1)
+
+
+class Model_Eager(Model, tf.contrib.eager.Network):
```
Comment (rajivhockey97) [inline cleverhans/model_eager.py]: tfe.Network is deprecated in tensorflow rc 1.9.0. We could inherit from tf.keras.Model instead.

File: cleverhans/model.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
 from abc import ABCMeta
+import tensorflow as tf
```
Comment (david-berthelot) [inline cleverhans/model.py]: New import but not used.

File: cleverhans/model.py [hunk 3]
```diff
@@ -101,7 +102,6 @@ def __init__(self, callable_fn, output_layer):
         :param output_layer: A string of the output layer returned by the
                              function. (Usually either "probs" or "logits".)
         """
-
```
Comment (david-berthelot) [inline cleverhans/model.py]: This change is unrelated to eager.

File: cleverhans/attacks_eager.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+    exit(1)
+
+try:
+    from cleverhans.model_eager import Model_Eager
+except ImportError:
+    error = 'Import Error: cleverhans.model_eager'
+    raise ImportError(error)
+
+
+class Attack(object):
```
Comment (david-berthelot) [inline cleverhans/attacks_eager.py]: I don't think you should redefine the base class Attack. May be it's an opportunity to refactor the base class from attacks.py so that you can inherit from it to form a new class (say AttackTFE).

File: cleverhans/attacks_eager.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+    error = ('For eager execution',
+             'use Tensorflow version greather than 1.8.0.')
+    print(error)
+    exit(1)
+
+try:
+    from cleverhans.model_eager import Model_Eager
+except ImportError:
+    error = 'Import Error: cleverhans.model_eager'
+    raise ImportError(error)
```
Comment (david-berthelot) [inline cleverhans/attacks_eager.py]: Just for my understanding, why do you intercept the import error if you're reraising it? Is it only to modify the error message? For what reason?

File: cleverhans/attacks_eager.py [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+
+    def __init__(self, model, back='tfe', dtypestr='float32'):
+        """
+        :param model: An instance of the cleverhans.model.Model_Eager class.
+        :param back: The backend to use. Currently 'tfe' is the only option.
+        :param dtypestr: datatype of the input data samples and crafted
+                        adversarial attacks.
+        """
+        # Validate the input arguments.
+        if not(back == 'tfe'):
```
Comment (david-berthelot) [inline cleverhans/attacks_eager.py]: A simpler way to do it is
if back != 'tfe':

File: cleverhans/attacks_eager.py [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+            raise ValueError("Backend argument must be 'tfe'.")
+        if back == 'tfe':
```
Comment (david-berthelot) [inline cleverhans/attacks_eager.py]: Isn't this if statement always true?

File: cleverhans/attacks_eager.py [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+                                    logits=logits, labels=labels)
+        except AttributeError:
+            out = tf.nn.softmax_cross_entropy_with_logits(
+                                    logits=logits, labels=labels)
+        if mean:
+            out = tf.reduce_mean(out)
+        return out
+
+
+class ListOfAttacks(object):
```
Comment (david-berthelot) [inline cleverhans/attacks_eager.py]: 1. What is the use case for this?
2. Why use a class instead of say?
ATTACKS = {'fsgm': FastGradientMethod}

File: cleverhans/attacks_eager.py (hunk 5 above)
Comment (feedforward) [reply cleverhans/attacks_eager.py]: tensorflow version (require >= 1.8.0) is to checked before importing Model_Eager, 
but PEP8 requires all import statements to be at beginning of the file.

i used try ... except as a hack around this deadlock, if there is an elegant solution please let me know.

File: cleverhans/attacks_eager.py (hunk 6 above)
Comment (feedforward) [reply cleverhans/attacks_eager.py]: True, will update PR to reflect this change.

File: cleverhans/attacks_eager.py (hunk 7 above)
Comment (feedforward) [reply cleverhans/attacks_eager.py]: True, redundant snippet is used to stay similar to previous code in attacks.py 
`        if not(back == 'tf'):
            raise ValueError("Backend argument must be 'tf'.")

        if back == 'tf':
            import tensorflow as tf
            self.tf_dtype = tf.as_dtype(dtypestr)`
strictly speaking this is not necessary, let me know what you think.

File: cleverhans/attacks_eager.py (hunk 8 above)
Comment (feedforward) [reply cleverhans/attacks_eager.py]: In cleverhans_tutorials for tf backend each attack has one tutorial, 
but for eager backend we can have one tutorial that accepts --attack as a argument.
mnist_tutorial_eager.py expects  a string for --attack argument, (eg: 'fgsm') and maps it to Attack class 
 (eg: FastGradientMethod).

This class is used for above mapping attack name (string) to Attack class, to reduce code repetition, what are your thoughts on this ?

File: cleverhans/model.py (hunk 2 above)
Comment (feedforward) [reply cleverhans/model.py]: True, updated PR will reflect the change.

File: cleverhans/model.py (hunk 3 above)
Comment (feedforward) [reply cleverhans/model.py]: True, updated PR will reflect the change.

File: cleverhans/model_eager.py (hunk 1 above)
Comment (feedforward) [reply cleverhans/model_eager.py]: #430 , suggestions were in favour of not using the keras dependencies.

File: cleverhans/attacks_eager.py (hunk 4 above)
Comment (feedforward) [reply cleverhans/attacks_eager.py]: I agree, 
few options i could think of, let me know your thoughts on which one looks best .... 
Option 1:
base class Attack uses feed_dict and runs through session, inheriting AttackTFE from Attack will bring graph based code structure to AttackTFE, but we can overload the functions, 

PRO: this way no changes are required to existing code base,  
CONS: unnecessary graph def code inherited to AttackTFE, not elegant.

Option 2:
Add a new base class Attack_BASE to contain functions common to Attack and AttackTFE, then inherit both these classes from Attack_BASE.
PRO: hierarchical structure is maintained.
CON: few changes to the existing code base.

**Comment by feedforward** [comment]
@david-berthelot  requested changes are applied, with replies to the your comments, let me know what your thoughts are, particularly on handling AttackTFE class

File: cleverhans/attacks_eager.py (hunk 5 above)
Comment (david-berthelot) [reply cleverhans/attacks_eager.py]: Shouldn't it raise an error if tf version is too old and eager is not present?
What's wrong with just:
```
from cleverhans.model_eager import Model_Eager
```
If it's present, it gets imported and if it's not present it will raise an import error.

File: cleverhans/attacks_eager.py (hunk 7 above)
Comment (david-berthelot) [reply cleverhans/attacks_eager.py]: I don't know I didn't review the previous code but it seems redundant just the same. No need to clean all existing code but writing new code you can fix stuff that you notice along the way.

File: cleverhans/attacks_eager.py (hunk 8 above)
Comment (david-berthelot) [reply cleverhans/attacks_eager.py]: Right but a dictionary is just as easy to import isn't it?
If the list is for the tutorial, may be move it to the tutorial.
Another problem with the list, is that it has to be manually maintained, someone adding a new attack could forget to put it in the list, although if it's just in the tutorial folder it's okay.

File: cleverhans/utils_eager.py [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+    x_shape = tf.shape(x)
+    x = tf.contrib.layers.flatten(x)
+    x /= (epsilon + tf.reduce_max(tf.abs(x), 1, keep_dims=True))
+    square_sum = tf.reduce_sum(tf.square(x), 1, keep_dims=True)
+    x_inv_norm = tf.rsqrt(np.sqrt(epsilon) + square_sum)
+    x_norm = tf.multiply(x, x_inv_norm)
+    return tf.reshape(x_norm, x_shape, scope)
+
+
+def kl_with_logits(p_logits, q_logits):
```
Comment (david-berthelot) [inline cleverhans/utils_eager.py]: What is specific to eager in this function?

File: cleverhans/utils_eager.py [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+    :param epsilon: stabilizes division
+    :return: the batch of l2 normalized vector
+    """
+    x_shape = tf.shape(x)
+    x = tf.contrib.layers.flatten(x)
+    x /= (epsilon + tf.reduce_max(tf.abs(x), 1, keep_dims=True))
+    square_sum = tf.reduce_sum(tf.square(x), 1, keep_dims=True)
+    x_inv_norm = tf.rsqrt(np.sqrt(epsilon) + square_sum)
+    x_norm = tf.multiply(x, x_inv_norm)
+    return tf.reshape(x_norm, x_shape, scope)
```
Comment (david-berthelot) [inline cleverhans/utils_eager.py]: I am confused by this function:
1. what is eager specific or should it just be a tf util?
2. I don't understand how this L2 batch normalization (of vectors I assume)
- you first divide x by max(abs(x_i)) which doesn't seem be related to L2 normalization.
- also why take the sqrt of epsilon, I am not saying this wrong, but I don't see the reason.
- you can simplify the code x_norm = x * x_inv_norm

File: cleverhans/attacks_eager.py (hunk 5 above)
Comment (feedforward) [reply cleverhans/attacks_eager.py]: True, above code is redundant now, will modify it in updated PR.

File: cleverhans/attacks_eager.py (hunk 7 above)
Comment (feedforward) [reply cleverhans/attacks_eager.py]: Sure, will make the changes in updated PR.

File: cleverhans/attacks_eager.py (hunk 8 above)
Comment (feedforward) [reply cleverhans/attacks_eager.py]: True, i will move this class to tutorial, 
dictionary is just as easy to import and serves the purpose, modified in the updated PR.

File: cleverhans/utils_eager.py (hunk 10 above)
Comment (feedforward) [reply cleverhans/utils_eager.py]: True, there is nothing eager specific to this function, haven't noticed that before. 
removed it.

File: cleverhans/utils_eager.py (hunk 9 above)
Comment (feedforward) [reply cleverhans/utils_eager.py]: SAME as previous comment.

File: cleverhans/attacks_eager.py [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+        self.tf_dtype = tf.as_dtype(dtypestr)
+        self.np_dtype = np.dtype(dtypestr)
+
+        if not isinstance(model, Model_Eager):
+            raise ValueError("The model argument should be an instance of"
+                             " the cleverhans.model.Model_Eager class.")
+        # Prepare attributes
+        self.layer_names = []
+
+    def get_or_guess_labels(self, x, kwargs):
```
Comment (david-berthelot) [inline cleverhans/attacks_eager.py]: What is specific to eager in this method? It looks very similar to the one in tf_attacks.py

File: cleverhans/attacks_eager.py [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+        :param x_val: A NumPy array with the original inputs.
+        :param **kwargs: optional parameters used by child classes.
+        :return: A NumPy array holding the adversarial examples.
+        """
+        tfe = tf.contrib.eager
+        x = tfe.Variable(x_val)
+        adv_x = self.generate(x, **kwargs)
+        return adv_x.numpy()
+
+    def clip_eta(self, eta, ord, eps):
```
Comment (david-berthelot) [inline cleverhans/attacks_eager.py]: What is specific to eager in this method? It looks very similar to the one in tf_attacks.py

File: cleverhans/attacks_eager.py [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+        Create a FastGradientMethod instance.
+        Note: the model parameter should be an instance of the
+        cleverhans.model.Model abstraction provided by CleverHans.
+        """
+        if not isinstance(model, Model_Eager):
+            model = CallableModelWrapper(model, 'probs')
+
+        super(FastGradientMethod, self).__init__(model, dtypestr)
+
+    def generate(self, x, **kwargs):
```
Comment (david-berthelot) [inline cleverhans/attacks_eager.py]: What is specific to eager in this method? It looks very similar to the one in tf_attacks.py

File: cleverhans/attacks_eager.py [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+                            Labels should be one-hot-encoded.
+        :param clip_min: (optional float) Minimum input component value
+        :param clip_max: (optional float) Maximum input component value
+        """
+        # Parse and save attack-specific parameters
+        assert self.parse_params(**kwargs)
+        labels, nb_classes = self.get_or_guess_labels(x, kwargs)
+        return self.fgm(x, labels=labels, targeted=(self.y_target is not None))
+
+    def parse_params(self, eps=0.3, ord=np.inf, y=None, y_target=None,
```
Comment (david-berthelot) [inline cleverhans/attacks_eager.py]: What is specific to eager in this method? It looks very similar to the one in tf_attacks.py

File: cleverhans/attacks_eager.py [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+        self.clip_max = clip_max
+
+        if self.y is not None and self.y_target is not None:
+            raise ValueError("Must not set both y and y_target")
+        # Check if order of the norm is acceptable given current implementation
+        if self.ord not in [np.inf, int(1), int(2)]:
+            raise ValueError("Norm order must be either np.inf, 1, or 2.")
+        return True
+
+    def fgm(self, x, labels, targeted=False):
```
Comment (david-berthelot) [inline cleverhans/attacks_eager.py]: Seems like the only eager specific code, so why not inherit from the tf Attack and just redefine this function.
You could also overload the create graph one to return an error since I believe you don't create on in eager.

File: cleverhans/utils_eager.py [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+import tensorflow as tf
+import time
+import warnings
+
+from .utils import batch_indices, _ArgsWrapper, create_logger
+
+_logger = create_logger("cleverhans.utils.eager")
+
+
+def model_loss(logits, labels, mean=True):
```
Comment (david-berthelot) [inline cleverhans/utils_eager.py]: What is specific to eager in this method? It looks very similar to the one in utils.tf

File: cleverhans/utils_eager.py [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+    except AttributeError:
+        out = tf.nn.softmax_cross_entropy_with_logits(
+                                                logits=logits, labels=labels)
+
+    if mean:
+        out = tf.reduce_mean(out)
+    return out
+
+
+def model_train(model, X_train=None, Y_train=None, save=False,
```
Comment (david-berthelot) [inline cleverhans/utils_eager.py]: This function follows the deprecated mode, it does not use a loss. Check the function train, which replaces it.

File: cleverhans_tutorials/mnist_tutorial_eager.py [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+
+FLAGS = flags.FLAGS
+
+
+# Keeps track of implemented attacks.
+# Maps attack string taken bash to attack class
+# -- 'fgsm' : FastGradientMethod
+
+ListOfAttacks = {
+                'fgsm': FastGradientMethod
```
Comment (david-berthelot) [inline cleverhans_tutorials/mnist_tutorial_eager.py]: This doesn't seem to follow PEP8 alignment, or does it?

File: cleverhans_tutorials/tutorial_models_eager.py [hunk 19]
```diff
@@ -0,0 +1,214 @@
+"""
```
Comment (david-berthelot) [inline cleverhans_tutorials/tutorial_models_eager.py]: We moved away from custom layers in tutorials, I believe the tf.layers work with tf.eager by default. You can simplify this file a lot by using them.

File: cleverhans/attacks_eager.py (hunk 11 above)
Comment (feedforward) [reply cleverhans/attacks_eager.py]: I see your point, do you suggest overloading Attack class definitions ?? (i'm open to this ... )
my thoughts -- 
in attacks_tf.py all the functions are defined as part of graph execution and run through session.
in eager mode they run on the fly.
code logic is similar by design, (they are meant to do same thing in different context.)

File: cleverhans/attacks_eager.py (hunk 12 above)
Comment (feedforward) [reply cleverhans/attacks_eager.py]: Same as before, lets discuss in the previous comment.

File: cleverhans/attacks_eager.py (hunk 13 above)
Comment (feedforward) [reply cleverhans/attacks_eager.py]: Same as before, lets discuss in the previous comment.

File: cleverhans/attacks_eager.py (hunk 14 above)
Comment (feedforward) [reply cleverhans/attacks_eager.py]: Same as before, lets discuss in the previous comment.

File: cleverhans/attacks_eager.py (hunk 15 above)
Comment (feedforward) [reply cleverhans/attacks_eager.py]: True, this can be one way to do it, let's finalise the approach in previous comment, then i will submit an updated PR.

File: cleverhans/utils_eager.py (hunk 17 above)
Comment (feedforward) [reply cleverhans/utils_eager.py]: Ohh, this change was done, after this PR is submitted, that is not an issue, i can update this accordingly.

File: cleverhans/utils_eager.py (hunk 16 above)
Comment (feedforward) [reply cleverhans/utils_eager.py]: True, will import it.

File: cleverhans_tutorials/mnist_tutorial_eager.py (hunk 18 above)
Comment (feedforward) [reply cleverhans_tutorials/mnist_tutorial_eager.py]: Looks misaligned, Not sure, it did pass the travis test though

File: cleverhans_tutorials/tutorial_models_eager.py (hunk 19 above)
Comment (feedforward) [reply cleverhans_tutorials/tutorial_models_eager.py]: above change was done after this PR is submitted, i will update this in next PR.

File: cleverhans/attacks_eager.py (hunk 11 above)
Comment (david-berthelot) [reply cleverhans/attacks_eager.py]: Yes, that's exactly my thoughts, the code is the same except for gradient the tape and execution, so inheriting from tf_attack so that the eager code is minimal is a much more impressive demonstration of its simplicity than having a lot of duplicate which makes more intimidating.

File: cleverhans_tutorials/tutorial_models_eager.py (hunk 19 above)
Comment (david-berthelot) [reply cleverhans_tutorials/tutorial_models_eager.py]: I'm not sure what's best here. I'll say check with Nicolas / Ian and see what they think. I personally think it's best to have a PR with all the up-to-date code rather than pushing obsolete code at commit time.

**Comment by david-berthelot** [comment]
A few more changes and it'll be done (the refactor of attack to reuse tf as much as possible since it's meant to be the same). Also need 3rd opinion on committing obsolete / deprecated code.

**Comment by goodfeli** [comment]
Re deprecated code: generally we should not add deprecated code to the main repo. The idea of deprecated code is that 3rd party users can continue using it for a while, the library may not immediately *remove* uses of deprecated code, but the library shouldn't ever *add* uses of deprecated code.

File: cleverhans/attacks_eager.py (hunk 11 above)
Comment (feedforward) [reply cleverhans/attacks_eager.py]: Sure, makes sense, will do that ...

**Comment by feedforward** [comment]
Refactored to
1. Inhereted from Attack class.
2. Uses recently merged loss class
3. Uses network similar to recently merged MakeBasicCNN

**Comment by feedforward** [comment]
Refactored eager code.

File: cleverhans/loss.py [hunk 20]
```diff
@@ -55,7 +55,7 @@ def fprop(self, x, y, **kwargs):
             x = x, self.attack(x)
         else:
             x = x,
-        y -= self.smoothing * (y - 1. / tf.cast(y.shape[-1], tf.float32))
+        #y -= self.smoothing * (y - 1. / tf.cast(y.shape[-1], tf.float32))
```
Comment (david-berthelot) [inline cleverhans/loss.py]: Did you comment this code by accident?

File: cleverhans/model.py [hunk 21]
```diff
@@ -68,8 +68,12 @@ def get_params(self):
         Provides access to the model's parameters.
         :return: A list of all Variables defining the model parameters.
         """
-        scope_vars = tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES,
-                                       self.scope)
+        if tf.executing_eagerly() == False:
```
Comment (david-berthelot) [inline cleverhans/model.py]: if tf.executing_eagerly() == False:

Simpler to write:
if not tf.executing_eagerly():

Or even better:
if tf.executing_eagerly():
   # raise
else:
  scope_vars

File: cleverhans/model.py [hunk 22]
```diff
@@ -68,8 +68,12 @@ def get_params(self):
...
+            scope_vars = tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES,
+                                           self.scope)
+        else:
+            error = 'For Eager execution - get_params must be overridden.'
+            raise NotImplementedError()
```
Comment (david-berthelot) [inline cleverhans/model.py]: Did you forget to put the error in the exception?

File: cleverhans/utils_tfe.py [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+                 and 'filename'
+    :param rng: Instance of numpy.random.RandomState
+    :param var_list: List of variables to train.
+    :param attack: Instance of the class cleverhans.attacks.attacks_eager
+    :param attack_args: Parameters required for the attack.
+    :return: True if model trained
+    """
+    args = _ArgsWrapper(args or {})
+    if ((attack is None) != (attack_args is None)):
+        error = "attack and attack_args must be passed together."
```
Comment (david-berthelot) [inline cleverhans/utils_tfe.py]: Why use an intermediate "error" variable?

File: cleverhans/model.py (hunk 22 above)
Comment (david-berthelot) [reply cleverhans/model.py]: In addition, I don't feel the intermediate variable "error" is needed here.

File: cleverhans/utils_tfe.py [hunk 24]
```diff
@@ -0,0 +1,232 @@
...
+    :param args: dict or argparse `Namespace` object.
+                 Should contain `batch_size`
+    :param attack: instance of the class cleverhans.attacks.attacks_eager
+    :param attack_args: parameters required for the attack.
+    :return: a float with the accuracy value
+    """
+    args = _ArgsWrapper(args or {})
+
+    if ((attack is None) != (attack_args is None)):
+        error = "attack and attack_args must be passed together."
```
Comment (david-berthelot) [inline cleverhans/utils_tfe.py]: Same comment about intermediate variable, it looks simpler to just pass the string to ValueError.

File: cleverhans/utils_tfe.py [hunk 25]
```diff
@@ -0,0 +1,232 @@
...
+                      taken from FLAGS.train_dir and FLAGS.filename
+    :return:
+    """
+    assert file_path != None, 'pretrained model ckpt is not given.' 
+    saver = tf.train.Saver()
+    saver.restore(file_path)
+    return True
+
+
+def model_argmax(model, samples,):
```
Comment (david-berthelot) [inline cleverhans/utils_tfe.py]: Trailing comma.

File: cleverhans_tutorials/mnist_tutorial_tfe.py [hunk 26]
```diff
@@ -0,0 +1,253 @@
...
+tfe = tf.contrib.eager
+
+FLAGS = flags.FLAGS
+
+
+# Keeps track of implemented attacks.
+# Maps attack string taken from bash to attack class
+# -- 'fgsm' : FastGradientMethod
+
+ListOfAttacks = {
```
Comment (david-berthelot) [inline cleverhans_tutorials/mnist_tutorial_tfe.py]: 1. It is confusing to call a dictionary ListOfAttacks.
2. It's a global constant, so it should be all uppercase.

How about AVAILABLE_ATTACKS?

File: cleverhans_tutorials/mnist_tutorial_tfe.py [hunk 27]
```diff
@@ -0,0 +1,253 @@
...
+    :param attack_string: adversarial attack name in string format
+    :return: attack class defined in cleverhans.attacks_eager
+    """
+
+    # List of Implemented attacks
+    attacks_list = ListOfAttacks.keys()
+
+    # Checking for  requested attack in list of available attacks.
+    if attack_string is None:
+        error = 'Attack type is not specified, list of available attacks\t'
```
Comment (david-berthelot) [inline cleverhans_tutorials/mnist_tutorial_tfe.py]: Same, it's simpler to just put the string directly in the exception line.

File: cleverhans_tutorials/mnist_tutorial_tfe.py [hunk 28]
```diff
@@ -0,0 +1,253 @@
...
+        error += attacks_list
+        raise AttributeError(error)
+    if attack_string not in attacks_list:
+        error = 'Attack not available, list of available attacks\t'
```
Comment (david-berthelot) [inline cleverhans_tutorials/mnist_tutorial_tfe.py]: Same, it's simpler to just put the string directly in the exception line.

File: cleverhans_tutorials/mnist_tutorial_tfe.py [hunk 29]
```diff
@@ -0,0 +1,253 @@
...
+        'nb_epochs': nb_epochs,
+        'batch_size': batch_size,
+        'learning_rate': learning_rate
+    }
+
+    # Initialize the attack object
+    attack_class = attack_selection(attack_string)
+    attack_params = {
+        'eps': 0.3, 'clip_min': 0.,
+        'clip_max': 1.}
```
Comment (david-berthelot) [inline cleverhans_tutorials/mnist_tutorial_tfe.py]: Why put this on an extra line when you still have space on the previous one?

File: cleverhans/utils_tfe.py (hunk 25 above)
Comment (feedforward) [reply cleverhans/utils_tfe.py]: True, removed.

File: cleverhans_tutorials/mnist_tutorial_tfe.py (hunk 27 above)
Comment (feedforward) [reply cleverhans_tutorials/mnist_tutorial_tfe.py]: same as above.

File: cleverhans_tutorials/mnist_tutorial_tfe.py (hunk 29 above)
Comment (feedforward) [reply cleverhans_tutorials/mnist_tutorial_tfe.py]: True, removed the extra line.

**Comment by goodfeli** [comment]
is this ready for re-review?

File: cleverhans/loss.py [hunk 30]
```diff
@@ -55,7 +55,14 @@ def fprop(self, x, y, **kwargs):
             x = x, self.attack(x)
         else:
             x = x,
-        #y -= self.smoothing * (y - 1. / tf.cast(y.shape[-1], tf.float32))
+
+        # Catching RuntimeError: Variable -= value not supported.
```
Comment (david-berthelot) [inline cleverhans/loss.py]: Not supported "by tf.eager".

File: cleverhans_tutorials/mnist_tutorial_tfe.py [hunk 31]
```diff
@@ -50,7 +48,7 @@
 # Maps attack string taken from bash to attack class
 # -- 'fgsm' : FastGradientMethod
 
-ListOfAttacks = {
+AVAILABLE_ATTACKS = {
                  'fgsm': FastGradientMethodTFE
```
Comment (david-berthelot) [inline cleverhans_tutorials/mnist_tutorial_tfe.py]: The alignment looks strange here, should be 4 characters?

**Comment by feedforward** [comment]
The final changes are done. @goodfeli  Please have a look at this.

**Comment by feedforward** [comment]
@npapernot Hi, review process is done, @david-berthelot approved the final version, please merge this PR.
